src/main/java/HttpServerRunner.java

```diff
@@ -3,29 +3,33 @@
 
 public class HttpServerRunner {
 
-  HttpReaderWriter httpReaderWriter;
   HandlerRouter handlerRouter;
-
-  public HttpServerRunner(HttpReaderWriter httpReaderWriter, HandlerRouter handlerRouter) {
-    this.httpReaderWriter = httpReaderWriter;
+  IRequestBuilder httpRequestBuilder;
+  HttpResponseWriter httpResponseWriter;
+    public HttpServerRunner(HandlerRouter handlerRouter, IRequestBuilder httpRequestBuilder, HttpResponseWriter httpResponseWriter) {
     this.handlerRouter = handlerRouter; 
+    this.httpRequestBuilder = httpRequestBuilder;
+    this.httpResponseWriter = httpResponseWriter;
   }
 
-  public void runServer() throws IOException {
-    HttpRequest request;
+  public void run(IMessageIO client) {
+   HttpRequest request;
     Ihandler handler;
     HttpResponse response;
-
-    while (true) {
-      try {
-        request = httpReaderWriter.getHttpRequest();
-        handler = handlerRouter.selectHandler(request);
-      }
-      catch (InvalidHttpRequestException e) {
-        handler = handlerRouter.selectHandlerBadRequest();
-      }
-      response = handler.generateResponse();
-      httpReaderWriter.sendHttpResponse(response);
-   }
+    try {
+      request = httpRequestBuilder.createRequest(client);
+      handler = handlerRouter.selectHandler(request);
+    }
+    catch (InvalidHttpRequestException e) {
+      handler = handlerRouter.selectHandlerBadRequest();
+    }
+    catch (IOException e) {
```

Comment:
Where it should be handled and how is up to you - but my point is that the reason it's being thrown here is because `httpRequestBuilder` or `handlerRouter` depend on something that throw it–they don't actually throw it themselves.

By letting it bubble up you're sort of exposing that and forcing `HttpServerRunner` to also depend on the "details" of those two classes it uses.